File: thorp_toolkit/python/thorp_toolkit/semantic_layer.py
import json

import rospy

from moveit_msgs.msg import CollisionObject
from shape_msgs.msg import SolidPrimitive
from thorp_msgs.msg import ThorpError
from thorp_msgs.srv import UpdateCollisionObjs

from geometry import pose2d2str, TF2
from singleton import Singleton


class SemanticLayer:
    """ Singleton providing a simplified interface for adding/removing objects to the costmaps semantic layer """
    __metaclass__ = Singleton

    def __init__(self):
        self._lcm_sl_srv = rospy.ServiceProxy('move_base_flex/local_costmap/semantic_layer/update_objects',
                                              UpdateCollisionObjs)
        self._gcm_sl_srv = rospy.ServiceProxy('move_base_flex/global_costmap/semantic_layer/update_objects',
                                              UpdateCollisionObjs)
        self._lcm_sl_srv.wait_for_service(30.0)
        self._gcm_sl_srv.wait_for_service(30.0)

    def add_object(self, name, type, pose, size, costmap='both'):
        """
        Add an semantic object to one or both costmaps. Pose will be transformed to map frame.
        :param name: object name
        :param type: object type (obstacle, free space, etc.)
        :param pose: object pose (PoseStamped)
        :param size: object size (3 floats list)
        :param costmap: 'local', 'global' or 'both'
        """
        sp = SolidPrimitive()
        sp.type = SolidPrimitive.BOX
        sp.dimensions = size
        co = CollisionObject()
        co.operation = CollisionObject.ADD
        co.id = name
        co.type.db = json.dumps({'table': 'NONE', 'type': type, 'name': name})
        co.header.frame_id = 'map'
        co.header.stamp = pose.header.stamp
        co.primitives.append(sp)
        co.primitive_poses.append(TF2().transform_pose(pose, pose.header.frame_id, co.header.frame_id).pose)
        try:
            if costmap in ['local', 'both']:
                resp = self._lcm_sl_srv([co])
                if resp.error.code != ThorpError.SUCCESS:
                    rospy.logerr("Unable to add object %s to %s costmap: %d", name, costmap, resp.error.code)
                    return False
            if costmap in ['global', 'both']:
                resp = self._gcm_sl_srv([co])
                if resp.error.code != ThorpError.SUCCESS:
                    rospy.logerr("Unable to add object %s to %s costmap: %d", name, costmap, resp.error.code)
                    return False
            rospy.loginfo("Added object %s of type %s at %s to %s costmap", name, type, pose2d2str(pose), costmap)
            return True
        except rospy.ServiceException as err:
            rospy.logerr("Unable to add object %s to %s costmap: %s", name, costmap, str(err))
            return False

    def remove_object(self, name, type, costmap='both'):
        """
        Remove an object from one or both costmaps.
        :param name: object name
        :param type: object type
        :param costmap: 'local', 'global' or 'both'
        """
        co = CollisionObject()
        co.operation = CollisionObject.REMOVE
        co.id = name
        co.type.db = json.dumps({'table': 'NONE', 'type': type, 'name': name})
        try:
            if costmap in ['local', 'both']:
                resp = self._lcm_sl_srv([co])
                if resp.error.code != ThorpError.SUCCESS:
                    rospy.logerr("Unable to remove object %s from %s costmap: %d", name, costmap, resp.error.code)
                    return False
            if costmap in ['global', 'both']:
                resp = self._gcm_sl_srv([co])
                if resp.error.code != ThorpError.SUCCESS:
                    rospy.logerr("Unable to remove object %s from %s costmap: %d", name, costmap, resp.error.code)
                    return False
            rospy.loginfo("Removed object %s of type %s from %s costmap", name, type, costmap)
            return True
        except rospy.ServiceException as err:
            rospy.logerr("Unable to remove object %s from %s costmap: %s", name, costmap, str(err))
            return False

File: thorp_toolkit/src/semantic_layer.cpp
#include "thorp_toolkit/semantic_layer.hpp"

#include <thorp_msgs/UpdateCollisionObjs.h>

#include <rr_geometry/converters.hpp>
namespace rrg = rapyuta::geometry_2d;

namespace rr::nav::semantic_layer
{
ServiceClient::ServiceClient()
{
  lcm_sl_srv_ =
      nh_.serviceClient<thorp_msgs::UpdateCollisionObjs>("move_base_flex/local_costmap/semantic/update_objects");
  gcm_sl_srv_ =
      nh_.serviceClient<thorp_msgs::UpdateCollisionObjs>("move_base_flex/global_costmap/semantic/update_objects");

  if (!lcm_sl_srv_.waitForExistence(ros::Duration(10.0)) || !gcm_sl_srv_.waitForExistence(ros::Duration(1.0)))
  {
    ROS_ERROR("Failed to find the semantic layer's update_objects service");
  }
}

bool ServiceClient::addObject(const std::string& name, const std::string& type, const geometry_msgs::PoseStamped& pose,
                              const geometry_msgs::Vector3& size, const std::string& costmap)
{
  rr_nav_semantic_layer::Object obj;
  obj.operation = rr_nav_semantic_layer::Object::ADD;
  obj.name = name;
  obj.type = type;
  obj.pose = pose;
  obj.dimensions = size;

  if ((costmap == "local" || costmap == "both") && !callSrv(lcm_sl_srv_, obj, "local"))
    return false;
  if ((costmap == "global" || costmap == "both") && !callSrv(gcm_sl_srv_, obj, "global"))
    return false;

  ROS_INFO("Added object %s of type %s at %s to %s costmap%s", name.c_str(), type.c_str(), rrg::toStr(pose).c_str(),
           costmap.c_str(), costmap == "both" ? "s" : "");
  return true;
}

bool ServiceClient::removeObject(const std::string& name, const std::string& type, const std::string& costmap)
{
  thorp_msgs::UpdateCollisionObjs::Request req;
  thorp_msgs::UpdateCollisionObjs::Response resp;

  rr_nav_semantic_layer::Object obj;
  obj.operation = rr_nav_semantic_layer::Object::REMOVE;
  obj.name = name;
  obj.type = type;

  if ((costmap == "local" || costmap == "both") && !callSrv(lcm_sl_srv_, obj, "local"))
    return false;
  if ((costmap == "global" || costmap == "both") && !callSrv(gcm_sl_srv_, obj, "global"))
    return false;

  ROS_INFO("Removed object %s of type %s from %s costmap%s", name.c_str(), type.c_str(), costmap.c_str(),
           costmap == "both" ? "s" : "");
  return true;
}

bool ServiceClient::callSrv(ros::ServiceClient& srv, const rr_nav_semantic_layer::Object& obj,
                            const std::string& costmap)
{
  thorp_msgs::UpdateCollisionObjs::Request req;
  thorp_msgs::UpdateCollisionObjs::Response resp;
  req.objects.push_back(obj);

  if (!srv.call(req, resp))
  {
    ROS_ERROR_STREAM("Semantic layer's update_objects service call failed");
    return false;
  }
  if (resp.code != thorp_msgs::UpdateCollisionObjs::Response::SUCCESS)
  {
    bool add = obj.operation == rr_nav_semantic_layer::Object::ADD;
    // clang-format off
    ROS_ERROR_STREAM((add ? "Add" : "Remove") << " object " << obj.name <<
                     (add ? " to " : " from ") << costmap << " costmap "
                     << "failed with error " << resp.code << ": " << resp.text);
    // clang-format on
    return false;
  }
  return true;
}

}  // namespace rr::nav::semantic_layer
